### ISSUE TITLE
Pod phase changed when container exits with 0 in 1.27

### ISSUE THREAD

**Issue by mimowo** [kubernetes/kubernetes#118310]
/sig node

After https://github.com/kubernetes/kubernetes/pull/115331 the phase of terminal pods killed with killPodFunc is determined based on the exit codes of the containers. This was discovered in the context of preemption while doing a cherry-pick to 1.26, and observing that the phase changed: https://github.com/kubernetes/kubernetes/pull/118221.

Summary of the phase transitions when the pod container exits with 0 (when restartPolicy=Always):
| scenario                                         | 1.26             |  1.27 
| -----------                                          | -----------       |  -----------
|  preemption                                  |  Failed         | Succeeded
|  eviction due to node pressure                      |  Failed         | Succeeded
| node shutdown                            | Failed | Succeeded
| active deadline exceeded           | Failed          | Failed

Note that eviction due to exceeding storage limits isn't affected because it uses grace period=0. In that case, SIGKILL is sent directly, and the containers end with 137. 

Are the changed phases a bug?
- if a bug, how to approach fixing it?
- if not a bug, we should clean up the setting of the phase as Failed in the component status updates, like here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/preemption/preemption.go#L106

Note: The change of phase for "node shutdown" was discovered earlier and uncovered a DaemonSet issue that was fixed: https://github.com/kubernetes/kubernetes/issues/117018

**Comment by mimowo** [comment]
FYI @smarterclayton @bobbypage @SergeyKanzhelev

**Comment by dchen1107** [comment]
This changed the semantics of the pod phase. The question is if we know how the controllers would act upon the change? At least we should take care all controllers offered by K8s out-of-box.

**Comment by mimowo** [comment]
One k8s controller is DaemonSet. It is fixed already under: https://github.com/kubernetes/kubernetes/pull/117073

**Comment by alculquicondor** [comment]
I remember that when we looked at #117018 we verified that this was the only controller that was affected.

**Comment by mimowo** [comment]
An update: I've checked in [WIP PR](https://github.com/kubernetes/kubernetes/pull/118345) if we just remove the explicit setting of phase=Failed in the evicting Kubelet components, then the reason & message fields aren'y set. So we need to either make a bigger change or keep it as is.

**Comment by HaveFun83** [comment]
any news here? We run 1.27.4 and stateful sets are still affected

**Comment by alculquicondor** [comment]
Can you provide more details? Maybe even some repro steps?

**Comment by mimowo** [comment]
Could you share the yaml for the stateful set? Best if you could open a dedicated issue with the minimal reproducible scenario. I suppose your containers intercept SIGTERM and exit with status 0, right?

**Comment by HaveFun83** [comment]
> Could you share the yaml for the stateful set? Best if you could open a dedicated issue with the minimal reproducible scenario. I suppose your containers intercept SIGTERM and exit with status 0, right?

We have GracefulNodeShutdown enabled
```
shutdownGracePeriod: 5m0s
shutdownGracePeriodCriticalPods: 1m40s
 ```
 Both statefulsets are managed by the offical helmcharts loki / minio
 and this happens after all controlplane nodes reboot near simultaneously
 
<details>
<summary>loki statefulset</summary>

```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app.kubernetes.io/instance: loki-system
    app.kubernetes.io/name: loki
    kustomize.toolkit.fluxcd.io/name: loki
    kustomize.toolkit.fluxcd.io/namespace: flux-system
  name: loki
  namespace: loki-system
spec:
  persistentVolumeClaimRetentionPolicy:
    whenDeleted: Retain
    whenScaled: Retain
  podManagementPolicy: OrderedReady
  replicas: 3
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/instance: loki-system
      app.kubernetes.io/name: loki
  serviceName: loki-headless
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/instance: loki-system
        app.kubernetes.io/name: loki
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchLabels:
                app.kubernetes.io/instance: loki-system
                app.kubernetes.io/name: loki
            namespaces:
            - loki-system
            topologyKey: kubernetes.io/hostname
      containers:
      - args:
        - -config.file=/etc/loki/config.yaml
        - -config.expand-env
        env:
        - name: S3_ACCESS_KEY_ID
          valueFrom:
            secretKeyRef:
              key: root-user
              name: minio-admin
        - name: S3_SECRET_ACCESS_KEY
          valueFrom:
            secretKeyRef:
              key: root-password
              name: minio-admin
        image: grafana/loki:2.8.3
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /ready
            port: http
            scheme: HTTP
          initialDelaySeconds: 45
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        name: loki
        ports:
        - containerPort: 3100
          name: http
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /ready
            port: http
            scheme: HTTP
          initialDelaySeconds: 45
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        securityContext:
          readOnlyRootFilesystem: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /tmp
          name: tmp
        - mountPath: /etc/loki
          name: config
        - mountPath: /loki
          name: storage
        - mountPath: /loki/rules-temp
          name: rules-temp
      dnsPolicy: ClusterFirst
      nodeSelector:
        node-role.kubernetes.io/control-plane: ""
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        fsGroup: 10001
        runAsGroup: 10001
        runAsNonRoot: true
        runAsUser: 10001
      serviceAccount: loki
      serviceAccountName: loki
      terminationGracePeriodSeconds: 4800
      tolerations:
      - effect: NoSchedule
        key: node-role.kubernetes.io/control-plane
      - effect: NoSchedule
        key: node-role.kubernetes.io/master
      volumes:
      - emptyDir:
          medium: Memory
          sizeLimit: 64Mi
        name: rules-temp
      - emptyDir:
          medium: Memory
        name: tmp
      - configMap:
          defaultMode: 420
          items:
          - key: config.yaml
            path: config.yaml
          name: loki-config-87t5k474m9
        name: config
  updateStrategy:
    type: RollingUpdate
  volumeClaimTemplates:
  - apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/instance: loki-system
        app.kubernetes.io/name: loki
      name: storage
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 2Gi
      storageClassName: local-controlplane-ssd
      volumeMode: Filesystem
    status:
      phase: Pending
```

</details>

**Comment by mimowo** [comment]
@HaveFun83 thanks, I have some additional questions: what is the reason for reboot of the node? Does it get replaced with a new node, but the same UID? It would also be helpful if you create a dedicated issue xreferencing this one and describe the minimal reproducible scenario. 

In the meanwhile I can reproduce it now as a minimal scenario using the following yaml on the Kind cluster:
```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: example
spec:
  selector:
    matchLabels:
      app: example
  serviceName: "example-svc"
  replicas: 3
  template:
    metadata:
      labels:
        app: example
    spec:
      containers:
        - name: busybox
          image: busybox
          command: ["sh"]
          args:
          - "-c"
          - |
            sleep 9999999 &
            PID=$!
            echo "Hello world!"
            _term () {
                kill $PID
                echo "Caught SIGTERM!"
            }
            trap _term SIGTERM
            wait $PID
            trap - TERM
            # Wait for the long running sleep to exit
            wait $PID
            exit 0
```
Then, I simulate the node shutdown, and all the pods get into the `Completed` phase, and are not deleted. As a consequences they are not recreated by the StatefulSet controller as it only recreates the Failed pods: https://github.com/kubernetes/kubernetes/blob/370c85f5ab0b0bfc3b30f235ddb040c246b0e1ff/pkg/controller/statefulset/stateful_set_control.go#L379. 

I can see a few options for solving this, but all might have tricky consequences:
1. Adjust the eviction procedure in the kubelet node shutdown manger to set the `deletionTimestamp`. IIUC this was also proposed as a solution in https://github.com/kubernetes/kubernetes/issues/116965#issuecomment-1667506435. 
2. Adjust the external controller to delete such pods during the node maintenance procedure. For example, if the nodes are recreated with different names, PodGC would clean up the completed nodes. If the pods are recreated but with different UIDs, then GCP controller manger (on GKE) would clean up the pods on the recreated nodes: https://github.com/kubernetes/cloud-provider-gcp/blob/master/cmd/gcp-controller-manager/node_csr_approver.go
3. Mark pods as Failed when shutting down a node, even if the exit code is 0 (as before https://github.com/kubernetes/kubernetes/pull/115331)
4. Adjust StatefulSet controller to recreate also Succeeded pods

These are in the order of my preference. It seems clean to reflect the termination by setting the `deletionTimestamp`. I suspect the decision was to save QPS in kubelet, but this was recently increased, so I suspect we could revisit that. 

FYI / WDYT @bobbypage @smarterclayton @alculquicondor @aojea ?

**Comment by alculquicondor** [comment]
Comments on the proposals:

1. How would the deletionTimestamp help? Would the statefulset controller replace pods that are Succeeded but have a deletionTimestamp?
2. IIUC, this is not a problem in GKE, given the existing GCP controller manager behavior. @bobbypage? In any case, I agree that Statefulsets should not return exit code 0, since they are meant to run indefinitely.
3. No, we already agreed that 0 should be treated as success.
4. Yes, this should be done, regardless of whatever else we do.

**Comment by mimowo** [comment]
> How would the deletionTimestamp help? Would the statefulset controller replace pods that are Succeeded but have a deletionTimestamp?

The deletionTimestamp would make the API server delete the pod once terminal. Then, stateful set would recreate it, because it would recognize there are not enough pods.

**Comment by mimowo** [comment]
> 4\. Yes, this should be done, regardless of whatever else we do.

After the second thought I like it, I was concerned about backwards compatibility. However, the StatefulSet requires `restartPolicy: Always`, so it makes sense to restart also Succeeded pods, since succeeded containers are already restarted at the kubelet level.

**Comment by alculquicondor** [comment]
For anyone thinking of submitting a PR to fix Statefulset, here's the fix we did for Daemonset https://github.com/kubernetes/kubernetes/pull/118716

**Comment by Kavantix** [comment]
I've also observed this for both Deployments and Statefulsets on 1.27.

Both had pods with status `Completed` while they were killed because the node was out of memory. After being killed they never started again even when there was memory available again.
Both also have a restart policy of 'Always'

In the status field of the completed pod of the statefulset there is the following:
```yaml
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2023-08-31T19:50:01Z"
    message: 'The node was low on resource: memory. Threshold quantity: 750Mi, available:
      756864Ki. Container *redacted* was using 10256Ki, request is 0, has
      larger consumption of memory. '
    reason: TerminationByKubelet
    status: "True"
    type: DisruptionTarget
  - lastProbeTime: null
    lastTransitionTime: "2023-08-22T14:53:41Z"
    reason: PodCompleted
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2023-08-31T19:50:01Z"
    reason: PodCompleted
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2023-08-31T19:50:01Z"
    reason: PodCompleted
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2023-08-22T14:53:41Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - *redacted*
    lastState:
      terminated:
        containerID: *redacted*
        exitCode: 137
        finishedAt: "2023-08-22T15:00:16Z"
        reason: Error
        startedAt: "2023-08-22T14:59:16Z"
    name: *redacted*
    ready: false
    restartCount: 6
    started: false
    state:
      terminated:
        containerID: *redacted*
        exitCode: 0
        finishedAt: "2023-08-31T19:50:00Z"
        reason: Completed
        startedAt: "2023-08-22T15:01:44Z"

```

**Comment by mimowo** [comment]
@Kavantix thanks for reporting the issue for Deployments. For statefulsets we already have a PR fixing them under review: https://github.com/kubernetes/kubernetes/pull/120398

**Comment by Kavantix** [comment]
@mimowo I was taking a look at the PR, and although it would restart the pods that are in the succeeded state I believe there is another problem that this would not fix.

Pods that got evicted got an `Evicted` state before, is getting back that state also part of this issue?

**Comment by Kavantix** [comment]
I've also identified a similar issue in a DaemonSet.
The pod is not set to `Completed` but it is set to `Error` when it is evicted because of memory issues.

The status field again has the same thing but the pod in this case did not exit with a statuscode of 0.
And again the pod is not restarted when memory pressure went away

```yaml
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2023-09-05T12:28:11Z"
    message: 'The node was low on resource: memory. Threshold quantity: 750Mi, available:
      766004Ki. '
    reason: TerminationByKubelet
    status: "True"
    type: DisruptionTarget
```

I am using kubernetes 1.27.3 (AKS) so I believe the fix for the daemon set should already be in there

**Comment by alculquicondor** [comment]
> After looking some more into it Deployment pods seem to be recreated but the Completed pods remain.

Terminated pods will be deleted eventually by the garbage collector (based on the configured threshold for terminated pods).

> I've also identified a similar issue in a DaemonSet.
The pod is not set to Completed but it is set to Error when it is evicted because of memory issues.
>
> The status field again has the same thing but the pod in this case did not exit with a statuscode of 0.
> And again the pod is not restarted when memory pressure went away

What is the `.status.phase`?

**Comment by Kavantix** [comment]
> What is the `.status.phase`?

The phase was `Failed` if I recall correclty. (Cannot check anymore)

**Comment by alculquicondor** [comment]
Could you open a separate issue with repro steps? preferably using a kind cluster.

**Comment by SergeyKanzhelev** [comment]
/kind feature

(for now unless we decide it is a bug)

**Comment by rouke-broersma** [comment]
> /kind feature
> 
> (for now unless we decide it is a bug)

This change causes pods to not be rescheduled when evicted due to 

> @Kavantix thanks for reporting the change impacts Deployments too. For statefulsets we already have a PR fixing them under review: #120398

I see the fix for statefulsets is listed for 1.29, any way we can have this backported as a bugfix since this is currently causing statefulset pods to not be recreated when evicted due to node pressure in our environments?

**Comment by mimowo** [comment]
> I see the fix for statefulsets is listed for 1.29, any way we can have this backported as a bugfix since this is currently causing statefulset pods to not be recreated when evicted due to node pressure in our environments?

This is the idea. We want to backport the fix down to 1.27. However, it is currently reverted: https://github.com/kubernetes/kubernetes/pull/120755 due to flakes that were exposed by the PR. FYI @aleksandra-malinowska

**Comment by mmontes11** [comment]
> It looks like the triggered test failure has been fixed. Can this be back-ported to 1.27 now?

This fix is quite important for the database operators space, as we heavily rely on the `StatefulSet` construct. See the following issue:
- https://github.com/mariadb-operator/mariadb-operator/issues/253

Could we please get it backported to 1.27? 🙏🏻

**Comment by mimowo** [comment]
> > It looks like the triggered test failure has been fixed. Can this be back-ported to 1.27 now?
> 
> This fix is quite important for the database operators space, as we heavily rely on the `StatefulSet` construct. See the following issue:
> 
> * [StatefulSet instances exit with exit code 0 when the `GracefulNodeShutdown` feature gate is used mariadb-operator/mariadb-operator#253](https://github.com/mariadb-operator/mariadb-operator/issues/253)
> 
> Could we please get it backported to 1.27? 🙏🏻

Yes, we want to cherry-pick, but first we need to revert of the fix revert https://github.com/kubernetes/kubernetes/pull/120755 :), because it exposed another bug and caused flakes.

We are working on fixing the bug here: https://github.com/kubernetes/kubernetes/pull/120731. Once this is done, we will re-revert the fix and backport to 1.27. FYI: @soltysh @aleksandra-malinowska @adilGhaffarDev

**Comment by mimowo** [comment]
> Could we please get it backported to 1.27? 🙏🏻

We have the cherry-picks pending approval, here for 1.27: https://github.com/kubernetes/kubernetes/pull/121921

**Comment by Chaunceyctx** [comment]
@mimowo  maybe another problem scenario ?
1. create job with 1 pod(use GPU or other device) and wait until pod becomes running
2. stop device plugin pod
3. restart kubelet, pod will be admitError because device plugin pod is down and cannot register **healthy devices** to kubelet
4. before https://github.com/kubernetes/kubernetes/pull/115331, pod phase will be Failed and JobController will create new pod to another node(device plugin pod is running). After https://github.com/kubernetes/kubernetes/pull/115331, pod phase may be Succeeded and jobController will never create new pod.

**Comment by mimowo** [comment]
> @mimowo maybe another problem scenario ?
> 
> 1. create job with 1 pod(use GPU or other device) and wait until pod becomes running

I'm not sure I'm following. I don't think https://github.com/kubernetes/kubernetes/pull/115331 changed the behavior for Job pods, because Jobs don't use `restartPolicy: Always` (see description in this issue). Previously Job pods would end up `Failed` or `Succeeded` (based on the container exit code). 

Also, in case of kubelet admission errors the phase is set `Failed` unconditionally: https://github.com/kubernetes/kubernetes/blob/c3ff08250eb434b9a4a51b0b23f9c3db91d5d1f5/pkg/kubelet/kubelet.go#L2273.

**Comment by Chaunceyctx** [comment]
@mimowo  LoL. so quick response! I am not sure. Aforesaid scenario is just my guess, I am finding code to prove it

**Comment by Chaunceyctx** [comment]
@mimowo so this issue is not a bug we need to fix?

**Comment by mimowo** [comment]
> @mimowo so this issue is not a bug we need to fix?

The issue is an umbrella to discuss potential bugs caused by the change in 1.27. So far we had two fixes: https://github.com/kubernetes/kubernetes/pull/117073, and https://github.com/kubernetes/kubernetes/pull/121389. I'm not aware of more bugs. 

Potentially there is still a small issue with Deployments, because the pods are not deleted (we may consider creating a follow up issue): https://github.com/kubernetes/kubernetes/issues/118310#issuecomment-1705055646 (they are deleted by StatefulSet for comparison). However, IMHO the priority is low as they are recreated (see https://github.com/kubernetes/kubernetes/issues/118310#issuecomment-1705218026) and gc-ed , (see https://github.com/kubernetes/kubernetes/issues/118310#issuecomment-1706620162).

**Comment by Chaunceyctx** [comment]
thanks!

> > @mimowo so this issue is not a bug we need to fix?
> 
> The issue is an umbrella to discuss potential bugs caused by the change in 1.27. So far we had two fixes: #117073, and #121389. I'm not aware of more bugs.
> 
> Potentially there is still a small issue with Deployments, because the pods are not deleted (we may consider creating a follow up issue): [#118310 (comment)](https://github.com/kubernetes/kubernetes/issues/118310#issuecomment-1705055646) (they are deleted by StatefulSet for comparison). However, IMHO the priority is low as they are recreated (see [#118310 (comment)](https://github.com/kubernetes/kubernetes/issues/118310#issuecomment-1705218026)) and gc-ed , (see [#118310 (comment)](https://github.com/kubernetes/kubernetes/issues/118310#issuecomment-1706620162)).

**Comment by alculquicondor** [comment]
/close

Let's reopen if we find another issue.

**Comment by buroa** [comment]
Currently seeing this on my Talos nodes with K8s v1.28.4. Seeing it happen on both Deployments and StatefulSets... maybe others. It happens during a graceful drain where pods get stuck in either a Failed or a Completed state. I am seeing Failed pods with exit code 143 and Completed pods with exit code 0.

I'm not sure this bug is fixed, either way.

**Comment by JJGadgets** [comment]
I’m seeing the same behaviour as @buroa on Talos 1.5.4 running Kubernetes 1.28.2 where a graceful node drain causes Deployment pods (I don’t currently have StatefulSets to test) to be stuck as Completed even after the drained node finishes starting up and is Ready status.

In addition to a 1 node graceful drain and reboot causing this behaviour, I did a full graceful reboot of my 3 node Talos cluster at the same time (knowing all pods will be down), and the same thing happens: when all nodes were back to Ready status, my 3 replicas Kyverno admission controller Deployment (with topologySpreadConstraints set to maxSkew 1, topologyKey hostname, and DoNotSchedule) were all stuck in Completed with no new pods being created by the Deployment, causing the mutating and validating webhooks from Kyverno to fail until I cleared the Completed pods manually.

**Comment by mimowo** [comment]
> Currently seeing this on my Talos nodes with K8s v1.28.4. Seeing it happen on both Deployments and StatefulSets... maybe others. It happens during a graceful drain where pods get stuck in either a Failed or a Completed state. I am seeing Failed pods with exit code 143 and Completed pods with exit code 0.

The bug for the recreation of Succeeded pods in StatefulSet is fixed in master https://github.com/kubernetes/kubernetes/pull/121389, is cherry-picked and it will be included in the next point releases, that is 1.27.9 and 1.28.5. The pods which are Failed (143) should have been recreated even before the fix.

> I’m seeing the same behaviour as @buroa on Talos 1.5.4 running Kubernetes 1.28.2 where a graceful node drain causes Deployment pods (I don’t currently have StatefulSets to test) to be stuck as Completed even after the drained node finishes starting up and is Ready status.

Could you please create a dedicated issue with the scenario described as precisely as possible, so that it is easy to reproduce? This will be very helpful for solving the issue.

**Comment by buroa** [comment]
> > Currently seeing this on my Talos nodes with K8s v1.28.4. Seeing it happen on both Deployments and StatefulSets... maybe others. It happens during a graceful drain where pods get stuck in either a Failed or a Completed state. I am seeing Failed pods with exit code 143 and Completed pods with exit code 0.
> 
> The bug for the recreation of Succeeded pods in StatefulSet is fixed in master #121389, is cherry-picked and it will be included in the next point releases, that is 1.27.9 and 1.28.5. The pods which are Failed (143) should have been recreated even before the fix.
> 
> > I’m seeing the same behaviour as @buroa on Talos 1.5.4 running Kubernetes 1.28.2 where a graceful node drain causes Deployment pods (I don’t currently have StatefulSets to test) to be stuck as Completed even after the drained node finishes starting up and is Ready status.
> 
> Could you please create a dedicated issue with the scenario described as precisely as possible, so that it is easy to reproduce? This will be very helpful for solving the issue.

@mimowo Here you go: https://github.com/kubernetes/kubernetes/issues/122122

**Comment by JJGadgets** [comment]
> Could you please create a dedicated issue with the scenario described as precisely as possible, so that it is easy to reproduce? This will be very helpful for solving the issue.

Will comment on #122122.
